### PR TITLE
[MM-53904] Fix MM-53904

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Fix MM-54904

#### Ticket Link
Fix https://mattermost.atlassian.net/browse/MM-53904

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: webapp/platform/client/src/client4.ts [hunk 1]
```diff
@@ -4221,6 +4222,32 @@ export default class Client4 {
...
+    private validateClient4URL(url?: string) {
+        const error = new ClientError(this.getUrl(), {
+            message: 'Invalid url.',
+            url,
+        })
+        if (!url) {
+            throw error;
+        }
+    
+        const path = url.split('?')[0].toLowerCase();
```
Comment (marianunez) [inline webapp/platform/client/src/client4.ts:4235]: Does this return `undefined` if there is no `?` ? Curious how this works with the test cases above

Comment (larkox) [reply webapp/platform/client/src/client4.ts:4235]: [String.split](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split) should always return a array of strings with at least one element. So...
- `"?foo".split('?')`===`["", "foo"]`
- `"foo".split('?')`===`["foo"]`
- `"".split('?')`===`[""]`

The only way to have an empty list according to the documentation is doing `"".split("")` because a slight change on semantics. Nevertheles, that will never be done in any case. So... it is safe to assume that `url.split('?')[0]` is always defined and a string.

Comment (marianunez) [reply webapp/platform/client/src/client4.ts:4235]: Ah yes, makes sense. Thanks for the detailed explanation!

File: webapp/platform/client/src/client4.test.ts [hunk 2]
```diff
@@ -40,6 +40,27 @@ describe('Client4', () => {
...
+                {url: 'something%2Fsomething%2F'},
+                {url: 'something%2fsomething%2f'},
+                {url: '../something'},
+                {url: 'something/./something'},
+                {url: 'something/%2e%2E'},
+                {url: 'something/%2E%2E/something'},
+                {url: 'something/%2e%2e/something'},
+                {url: 'something/.%2E'},
+                {url: '%2e./something'},
+            ]
```
Comment (marianunez) [inline webapp/platform/client/src/client4.test.ts:55]: 💯

**Review by marianunez — APPROVED** [review]
Thanks @larkox!

File: webapp/platform/client/src/client4.ts [hunk 3]
```diff
@@ -4221,6 +4222,32 @@ export default class Client4 {
             {method: 'post', body: JSON.stringify(body)},
         )
     }
+
+    private validateClient4URL(url?: string) {
```
Comment (hmhealey) [inline webapp/platform/client/src/client4.ts:4226]: Minor nit, could you rename this to `validateClientUrl`? The 4 followed by a capital letter looks a bit weird to me and our naming convention in the web app is to only capitalize the first letter of abbreviations

**Review by hmhealey — APPROVED** [review]
There's one thing I'd like changed, but otherwise, this looks good to me

File: webapp/platform/client/src/client4.ts [hunk 4]
```diff
@@ -4221,6 +4222,32 @@ export default class Client4 {
...
+        const path = url.split('?')[0].toLowerCase();
+        if (!path) {
+            throw error;
+        }
+    
+        if (path.includes('%2f')) {
+            throw error;
+        }
+    
+        const directories = path.split('/');
```
Comment (jupenur) [inline webapp/platform/client/src/client4.ts:4244]: Since this is the web, we'll also need to handle `\`. Browsers helpfully convert backslashes to forward slashes even though nobody really wants them to.

**Review by jupenur — CHANGES_REQUESTED** [review]

**Review by jupenur — CHANGES_REQUESTED** [review]
This whole approach of blocklisting specific bad sequences of characters is going to be very difficult to get right. For example I've just noticed that browsers strip at least `\n`, `\r`, and `\t` from URLs so that the URL path `/foo/.	./bar` is actually equivalent to `/foo/../bar`.

I would recommend switching to my original recommendation of validating inputs instead of attempting to sanitize the outgoing URL.

**Comment by hmhealey** [comment]
@jupenur We were trying to handle this on the Client4 side since it covers URLs that might come from other places.

That's annoying that there's other characters stripped from the URL like that. Do you have any reference for when that happens? I tried looking up the WhatWG spec for URLs but the only references I can find to those characters are how they're invalid in the domain, but I guess browsers aren't spec-compliant for `new URL` or `fetch`

**Comment by jupenur** [comment]
No this is weird browser-compat stuff that's not documented anywhere. That's what makes it difficult.

File: webapp/channels/src/components/team_controller/team_controller.tsx [hunk 5]
```diff
@@ -215,6 +215,7 @@ function TeamController(props: Props) {
 
     return (
         <Switch>
+            {/* /:team already validated on parent route */}
```
Comment (hmhealey) [inline webapp/channels/src/components/team_controller/team_controller.tsx:218]: 1/5, but we might want to repeat the pattern in more files so that people are more likely to copy the "correct" way of validating these

File: webapp/channels/src/utils/utils.tsx [hunk 6]
```diff
@@ -1333,15 +1333,15 @@ export function isValidPassword(password: string, passwordConfig: ReturnType<typ
 }
 
 function isChannelOrPermalink(link: string) {
-    let match = (/\/([^/]+)\/channels\/(\S+)/).exec(link);
+    let match = (/\/([a-z0-9\-_]+)\/channels\/([a-z0-9\-__]+)/).exec(link);
```
Comment (hmhealey) [inline webapp/channels/src/utils/utils.tsx:1336]: Were you going to revert these changes? I think these could also include a period as well

**Review by hmhealey — APPROVED** [review]

File: webapp/channels/src/utils/utils.tsx (hunk 6 above)
Comment (larkox) [reply webapp/channels/src/utils/utils.tsx:1336]: I have another PR that removes this completely. But you are right, the second part should also allow dots, since it could be a channel name. Will look into it.

File: webapp/channels/src/utils/path.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+export const TEAM_NAME_PATH_PATTERN = '[a-z0-9\\-_]+';
+
+// This should cover:
+// - Channel name
+// - Channel ID
+// - Group Channel Name (40 length UID)
+// - DM Name (userID__userID)
+// - Username prefixed by a @
+// - User ID
+// - Email
```
Comment (jupenur) [inline webapp/channels/src/utils/path.ts:18]: Where do we use emails with this? This doesn't seem to cover even the most common email address formats like `juho.nurminen+mattermost@example.com`. Should it?

**Review by jupenur — APPROVED** [review]
Overall this approach looks good, just one comment regarding emails. But if you decide to make any additional changes, please re-request a review.

File: webapp/channels/src/utils/path.ts (hunk 7 above)
Comment (larkox) [reply webapp/channels/src/utils/path.ts:18]: I would love input of @hmhealey on this, but according to the code, (and testing around that) you can open a DM by having a url like `/teamName/channels/userEmail` (if I remember correctly).

What I am not sure is whether we ever expose that kind of link anywhere, and if it is something we want to fully support, or support at all.

The reality is that emails are hard to deal with regular expressions, and we didn't want to go with the blacklisting approach, so not sure how to deal with this properly.

Comment (hmhealey) [reply webapp/channels/src/utils/path.ts:18]: I honestly don't remember when that was added. I think that and the ability to visit a DM by the user's ID might've been a feature for Shift, but I don't think they were ever documented or anything. I think Calls or something similar visits a DM by user ID, but I don't know of anything we have that accesses DMs by email.

Because I don't think anyone actually uses it, I'd be fine risking breaking that in certain cases. Since it's so hard to actually validate emails, I don't think it's worth trying to do that perfectly if we don't know people are actually using it

**Review by jupenur — APPROVED** [review]

**Review by hmhealey — APPROVED** [review]

**Comment by larkox** [comment]
@yasserfaraazkhan This requires more or less some smoke testing around navigation. Being able to properly go to different teams, different channels, go to the system console, go to the custom emoji modal, the integrations, etc... Anything that changes the URL.

**Review by yasserfaraazkhan — APPROVED** [review]
thank you @larkox ,
I've gone through creating teams, posting messages, attachments, premalinks, channels, gm, converting them, purchase, mfa, inviting users, adding members to channels , adding to teams.
did not see any error in the flows
